QuickFIXn/Transport/SslStreamFactory.cs

```diff
@@ -170,48 +170,56 @@ private bool VerifyRemoteCertificate(
+            var role = enhancedKeyUsage.Equals(CLIENT_AUTHENTICATION_OID, StringComparison.Ordinal) ? "client" : "server";
             _nonSessionLog.OnEvent(
                 $"Remote certificate is not intended for {role} authentication: It is missing enhanced key usage {enhancedKeyUsage}");
             return false;
         }
 
+        // If CA Certificate is specified then validate against the CA certificate, otherwise it is validated against the installed certificates
         if (string.IsNullOrEmpty(_socketSettings.CACertificatePath)) {
             _nonSessionLog.OnEvent("CACertificatePath is not specified");
-            return false;
-        }
-
-        string caCertPath = StringUtil.FixSlashes(_socketSettings.CACertificatePath);
-
-        // If CA Certificate is specified then validate against the CA certificate, otherwise it is validated against the installed certificates
-        X509Certificate2? cert = SslCertCache.LoadCertificate(caCertPath, null);
-        if (cert is null) {
-            _nonSessionLog.OnEvent(
-                $"Certificate '{caCertPath}' could not be loaded from store or path '{Directory.GetCurrentDirectory()}'");
-            return false;
-        }
-
-        X509Chain chain0 = new X509Chain();
-        chain0.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
-        // add all your extra certificate chain
-
-        chain0.ChainPolicy.ExtraStore.Add(cert);
-        chain0.ChainPolicy.VerificationFlags = X509VerificationFlags.AllowUnknownCertificateAuthority;
-        bool isValid = chain0.Build((X509Certificate2)certificate);
-
-        if (isValid)
-        {
-            // resets the sslPolicyErrors.RemoteCertificateChainErrors status
-            sslPolicyErrors &= ~SslPolicyErrors.RemoteCertificateChainErrors;
         }
         else
         {
-            sslPolicyErrors |= SslPolicyErrors.RemoteCertificateChainErrors;
+            string caCertPath = StringUtil.FixSlashes(_socketSettings.CACertificatePath);
+            
+            X509Certificate2? caCert = SslCertCache.LoadCertificate(caCertPath, null);
+            if (caCert is null) {
+                _nonSessionLog.OnEvent(
+                    $"Certificate '{caCertPath}' could not be loaded from store or path '{Directory.GetCurrentDirectory()}'");
+                return false;
+            }
+
+            X509Chain chain = new();
+            // add all your extra certificate chain
+            chain.ChainPolicy.TrustMode = X509ChainTrustMode.CustomRootTrust;
+            chain.ChainPolicy.CustomTrustStore.Add(caCert);
+            chain.ChainPolicy.RevocationMode = _socketSettings.CheckCertificateRevocation ? X509RevocationMode.Online : X509RevocationMode.NoCheck;
+
+            bool isValid = chain.Build((X509Certificate2)certificate);
+            if (isValid) {
+                bool isChainValid = true;
+                foreach (var status in chain.ChainStatus) {
```

Comment:
This part is probably unnecessary with the more restrictive CustomRootTrust but it also doesn't hurt to verify. This is really only involved for certs that don't go thru a Public CA.